teps/0044-decouple-task-composition-from-scheduling.md

```diff
@@ -0,0 +1,954 @@
+e.g. labels.
+
+### Runtime instead of authoring time
+
+These options pursue a solution that only works at runtime; this means Pipeline authors would not have any control
+over the scheduling.
+
+#### PipelineRun: emptyDir
+
+In this solution we use the values provided at runtime for workspaces to determine what to run. Specifically, we allow
+[`emptyDir`](https://github.com/tektoncd/pipeline/blob/a7ad683af52e3745887e6f9ed58750f682b4f07d/docs/workspaces.md#emptydir)
+to be provided as a workspace at the Pipeline level even when that workspace is used by multiple Tasks, and when that
+happens, we take that as the cue to schedule those Tasks together.
+
+For example given this Pipeline:
+
+```yaml
+kind: Pipeline
+metadata:
+  name: build-test-deploy
+spec:
+ workspaces:
+  - name: source-code
+  - name: test-results
+ tasks:
+ - name: get-source
+   workspaces:
+   - name: source-code
+     workspace: source-code
+   taskRef:
+     name: git-clone
+   params:
+   - name: url
+      value: $(params.url)
+    - name: revision
+      value: $(params.revision)
+ - name: run-unit-tests
+   runAfter: get-source
+   taskRef:
+     name: just-unit-tests
+   workspaces:
+   - name: source-code
+     workspcae: source-code
+   - name: test-results
+     workspace: test-results
+ - name: upload-results
+   runAfter: run-unit-tests
+   taskRef:
+     name: gcs-upload
+   params:
+   - name: location
+     value: gs://my-test-results-bucket/testrun-$(taskRun.name)
+   workspaces:
+   - name: data
+     workspace: test-results
+```
+
+Running with this PipelineRun would cause `get-source` and `run-unit-tests` to be run in one pod, with `upload-results`
+in another:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: run
+spec:
+  pipelineRef:
+    name: build-test-deply
+  workspaces:
+  - name: source-code
+    emptyDir: {}
+  - name: test-results
+    persistentVolumeClaim:
+      claimName: mypvc
+```
+
+Running with this PipelineRun would cause all of the Tasks to be run in one pod:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: run
+spec:
+  pipelineRef:
+    name: build-test-deply
+  workspaces:
+  - name: source-code
+    emptyDir: {}
+  - name: test-results
+    emptyDir: {}
+```
+
+Running with this PipelineRun would cause all of the Tasks to be run in separate pods:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: run
+spec:
+  pipelineRef:
+    name: build-test-deply
+  workspaces:
+  - name: source-code
+    persistentVolumeClaim:
+      claimName: otherpvc
+  - name: test-results
+    persistentVolumeClaim:
+      claimName: mypvc
+```
+
+Pros:
+* Allows runtime decisions about scheduling without changing the Pod
+
+Cons:
+* If it's important for a Pipeline to be executed in a certain way, that information will have to be encoded somewhere
```

Comment:
One con I have with this approach is that it is not that explicit. Using an emptyDir to say "run on the same pod", I would rather have a field that says that, explicitely.